addons/delivery/tests/test_delivery_cost.py

```diff
@@ -299,3 +300,88 @@ def test_01_taxes_on_delivery_cost(self):
         ])
 
         self.assertRecordValues(line, [{'price_subtotal': 9.09, 'price_total': 10.45}])
+
+    def test_delivery_cost_gift_card(self):
+        """
+        A customer has a carrier with the amount greater than the one to have
+        free shipping cost, then uses a gift card that lowers that amount to less
+        than the threshold: the shipping cost should still be 0.0
+        """
+
+        if "loyalty.card" not in self.env:
+            return
+
+        product_delivery_free = self.env['product.product'].create({
+            'name': 'Free Delivery Charges',
+            'type': 'service',
+            'list_price': 40.0,
+            'categ_id': self.env.ref('delivery.product_category_deliveries').id,
+        })
+        free_delivery = self.env['delivery.carrier'].create({
+            'name': 'Delivery Now Free Over 100',
+            'fixed_price': 40,
+            'delivery_type': 'fixed',
+            'product_id': product_delivery_free.id,
+            'free_over': True,
+            'amount': 100,
+        })
+        program_gift_card = self.env['loyalty.program'].create({
+            'name': 'Gift Cards',
+            'applies_on': 'future',
+            'program_type': 'gift_card',
+            'trigger': 'auto',
+            'reward_ids': [(0, 0, {
+                'reward_type': 'discount',
+                'discount': 1,
+                'discount_mode': 'per_point',
+                'discount_applicability': 'order',
+            })]
+        })
```

Comment:
`delivery` module doesn't depend on `loyalty` in any way.
If you install delivery and run its tests, this will fail.
It should be moved in `loyalty_delivery` module.